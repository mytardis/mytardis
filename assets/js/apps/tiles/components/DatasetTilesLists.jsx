import React, { Fragment, useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';
import Cookies from 'js-cookie';
import {
  fetchDatasetsForExperiment,
  fetchExperimentList,
  fetchExperimentPermissions,
  shareDataset,
} from './utils/FetchData';
import DatasetTiles from './DatasetTiles';
import ExperimentListDropDown from './SelectExperiment';
import DatasetPaneTopPanel from './DatasetPaneTopPanel';
import Spinner from '../../badges/components/utils/Spinner';

const DatasetTilesLists = ({ shareContainer, experimentId }) => {
  const [mainListData, setMainListData] = useState([]);
  const [shareListData, setShareListData] = useState([]);
  const [expListData, setExpListData] = useState();
  const [expListValue, setExpListValue] = useState();
  const [selectedDatasetIds, setSelectedDatasetIds] = useState([]);
  const [experimentPermissions, setExperimentPermissions] = useState({});
  const [mainListDataLoading, setMainListDataLoading] = useState(true);
  const csrfToken = Cookies.get('csrftoken');
  const spinnerCss = css`
    float: right;
  `;
  const onDragEnd = (result) => {
    // dropped nowhere
    if (!result.destination) {
      return;
    }
    // dropped in the same list
    if (result.destination.droppableId === result.source.droppableId) {
      return;
    }
    // dropped from share to main list
    if (result.destination.droppableId === 'main-list'
    && result.source.droppableId === 'share-list') {
      // get body data
      const data = shareListData[result.source.index];
      // update experiment dataset
      const datasetId = result.draggableId.split('_')[1];
      shareDataset(JSON.stringify(data), experimentId, datasetId)
        .then(() => {
          // fetch data and update main list
          fetchDatasetsForExperiment(experimentId)
            .then(listData => setMainListData(listData));
        });
    }
    // dropped from main to share list
    if (result.destination.droppableId === 'share-list'
    && result.source.droppableId === 'main-list') {
      // get body data
      const data = mainListData[result.source.index];
      // update experiment dataset
      const datasetId = result.draggableId.split('_')[1];
      shareDataset(JSON.stringify(data), expListValue, datasetId)
        .then(() => {
          // fetch data and update share list
          fetchDatasetsForExperiment(expListValue)
            .then(listData => setShareListData(listData));
        });
    }
  };
  const onDownloadSelect = (id, event) => {
    if (event.target.checked) {
      setSelectedDatasetIds([id].concat(selectedDatasetIds));
    } else {
      setSelectedDatasetIds(selectedDatasetIds.filter(item => item !== id));
    }
  };
  const onChange = (event) => {
    event.preventDefault();
    setExpListValue(event.target.value);
    fetchDatasetsForExperiment(event.target.value).then(result => setShareListData(result));
  };
  useEffect(() => {
    fetchDatasetsForExperiment(experimentId).then(result => setMainListData(result))
      .then(() => setMainListDataLoading(false));
    // load initial list
    fetchExperimentList().then((expList) => {
      // TODO
      setExpListData(expList);
      fetchDatasetsForExperiment(expList[0].id).then(result => setShareListData(result));
    });
    // load permissions
    fetchExperimentPermissions(experimentId).then(result => setExperimentPermissions(result));
  }, [experimentId]);
  return (
    <Fragment>
      {mainListDataLoading
        ? <Spinner override={spinnerCss} />
        : (
          <Fragment>
            <DatasetPaneTopPanel
              count={mainListData.length}
              experimentID={experimentId}
              selectedDatasets={selectedDatasetIds}
              csrfToken={csrfToken}
              experimentPermissions={experimentPermissions}
            />
            <DragDropContext onDragEnd={onDragEnd}>
              <Droppable droppableId="main-list">
                {provided => (
                  <div ref={provided.innerRef} {...provided.droppableProps}>
                    <DatasetTiles data={mainListData} listName="main-list" onDownloadSelect={onDownloadSelect} />
                    {provided.placeholder}
                  </div>
                )}
              </Droppable>
              {ReactDOM.createPortal(
                <Droppable droppableId="share-list">
                  {provided => (
                    <div ref={provided.innerRef} {...provided.droppableProps}>
                      {expListData
                        ? (
                          <ExperimentListDropDown
                            onChange={onChange}
                            experimentListData={expListData}
                          />
                        ) : <span />
              }
                      <DatasetTiles data={shareListData} listName="share-list" />
                      {provided.placeholder}
                    </div>
                  )}
                </Droppable>, shareContainer,
              )}
            </DragDropContext>
          </Fragment>
        )
        }
    </Fragment>
  );
};
DatasetTilesLists.propTypes = {
  shareContainer: PropTypes.object.isRequired,
  experimentId: PropTypes.string.isRequired,
};
export default DatasetTilesLists;
