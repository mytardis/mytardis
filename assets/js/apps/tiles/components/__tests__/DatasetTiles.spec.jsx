import React from 'react';
import { act } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import fetchMock from 'jest-fetch-mock';
import DatasetTilesLists from '../DatasetTilesLists';

const fakeData = [
  {
    id: 26,
    description: 'My demo dataset 25062019',
    directory: null,
    created_time: '2019-06-25 01:54:04.864000+00:00',
    modified_time: '2020-04-22 02:03:17.802951+00:00',
    immutable: false,
    instrument: null,
    experiments: [
      34,
    ],
    file_count: 1,
    url: '/dataset/26',
    size: 164495,
    size_human_readable: '160.6 KB',
    show_dataset_thumbnails: true,
  },
  {
    id: 30,
    description: 'New Dataset',
    directory: null,
    created_time: '2020-01-13 01:36:34.761466+00:00',
    modified_time: '2020-04-22 02:47:58.368048+00:00',
    immutable: false,
    instrument: null,
    experiments: [
      34,
    ],
    file_count: 1,
    url: '/dataset/30',
    size: 350565,
    size_human_readable: '342.3 KB',
    show_dataset_thumbnails: true,
  }];
const fakeExpList = [
  {
    id: 34,
    title: 'My Test Exp',
  },
  {
    id: 38,
    title: 'My Test Exp 2',
  }];

const fakeExpPermissions = {
  download_permissions: true,
  write_permissions: true,
  is_publication: false,
};

global.fetch = require('jest-fetch-mock');

configure({ adapter: new Adapter() });


let rootContainer = null;
let shareContainer = null;
let component = null;
let mainContainer = null;
beforeEach(async () => {
  fetchMock.mockResponse((request) => {
    if (request.url.endsWith('/dataset/')) {
      return Promise.resolve(JSON.stringify(fakeData));
    }
    if (request.url.endsWith('/experiment_list/')) {
      return Promise.resolve(JSON.stringify(fakeExpList));
    }
    if (request.url.endsWith('/get_experiment_permissions')) {
      return Promise.resolve(JSON.stringify(fakeExpPermissions));
    }
    return {
      status: 404,
      body: 'Not Found',
    };
  });

  rootContainer = global.document.createElement('div');
  rootContainer.setAttribute('id', 'root');
  mainContainer = global.document.createElement('div');
  mainContainer.setAttribute('id', 'datasets-pane');
  rootContainer.appendChild(mainContainer);

  shareContainer = global.document.createElement('div');
  shareContainer.setAttribute('id', 'experiment-tab-transfer-datasets');
  rootContainer.appendChild(shareContainer);
});
afterEach(() => {
  // cleanup on exiting

});
describe('render dataset tiles on page load', () => {
  it('should match snapshot', async () => {
    await act(async () => {
      component = mount(<DatasetTilesLists experimentId="1234" shareContainer={shareContainer} />, { attachTo: mainContainer });
    });
    component.update();
    expect(rootContainer).toMatchSnapshot();
  });
});
