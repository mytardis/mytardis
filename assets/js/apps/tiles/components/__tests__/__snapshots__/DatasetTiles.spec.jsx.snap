// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render dataset tiles on page load should match snapshot 1`] = `
<div
  id="root"
>
  <div
    id="datasets-pane"
  >
    <h3
      style="display: inline;"
    >
      2
       
      Datasets
    </h3>
    <a
      class="add-dataset btn btn-primary btn-sm pull-right"
      href="/experiment/1234/add-dataset"
      id="add-dataset"
    >
      <i
        class="fa fa-plus"
      />
      Add New
    </a>
    <br />
    <form
      action="/download/datafiles/"
      method="POST"
    >
      <input
        name="csrfmiddlewaretoken"
        type="hidden"
        value=""
      />
      <input
        name="expid"
        type="hidden"
        value="82"
      />
      <input
        name="comptype"
        type="hidden"
        value="tar"
      />
      <input
        name="organization"
        type="hidden"
        value="deep-storage"
      />
      <button
        class="btn btn-outline-secondary btn-sm download-selected"
        type="submit"
        value="submit"
      >
        <i
          class="fa fa-download"
        />
        Download Selected
      </button>
    </form>
    <p>
      <div
        class="form-horizontal"
      >
        <div
          class="form-group"
          style="margin: 0px; padding: 10px 0px 5px 10px; border: 1px solid #dddddd;"
        >
          <div
            class="col-md-12"
            style="margin-left: 0px; margin-bottom: 10px;"
          >
            <input
              autocomplete="off"
              class="input-xlarge"
              id="dataset-filter-text-{{cid}}"
              placeholder="Just start typing to filter datasets based on descriptions"
              style="width: 95%;"
              type="text"
            />
          </div>
        </div>
      </div>
    </p>
    <div
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="main-list"
    >
      <ul
        class="datasets thumbnails"
      >
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="main-list_26"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="main-list_26"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <li
            class="dataset card mb-2"
          >
            <div
              class="card-body"
            >
              <div
                class="float-left"
                style="margin-right: 10px;"
              >
                <div
                  style="display: flex; text-align: center; font-size: 32px;"
                >
                  <i
                    class="fa fa-folder-open o-6"
                  />
                </div>
                <br />
                <label
                  class="label label-info"
                  for="dataset-checkbox-26"
                  title="Mark dataset for download"
                >
                  <i
                    class="fa fa-download"
                  />
                  <input
                    class="dataset_checkbox ml-2"
                    id="dataset-checkbox-26"
                    name="dataset"
                    style="display: inline;"
                    type="checkbox"
                    value="26"
                  />
                </label>
              </div>
              <div
                class="float-left"
                style="margin-right: 10px;"
              />
              <div
                class="float-right"
                style="text-align: right;"
              >
                <p>
                  <span
                    class="badge badge-info"
                    title="1 file"
                  >
                    <i
                      class="fa fa-file"
                    />
                     
                    1
                  </span>
                </p>
                <p>
                  <span
                    class="badge badge-info"
                    title="Dataset size is"
                  >
                    160.64 KB
                  </span>
                </p>
              </div>
              <div
                class="float-left ds-thumb"
                style="margin-right: 10px;"
              >
                <canvas
                  style="width: 100px;"
                />
              </div>
              <div
                class="float-left"
              >
                <p
                  class="mb-2"
                >
                  <a
                    class="dataset-link"
                    href="/dataset/26"
                  >
                    My demo dataset 25062019
                  </a>
                </p>
                
                <div
                  style="clear: both;"
                />
              </div>
            </div>
          </li>
        </div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="main-list_30"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="main-list_30"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <li
            class="dataset card mb-2"
          >
            <div
              class="card-body"
            >
              <div
                class="float-left"
                style="margin-right: 10px;"
              >
                <div
                  style="display: flex; text-align: center; font-size: 32px;"
                >
                  <i
                    class="fa fa-folder-open o-6"
                  />
                </div>
                <br />
                <label
                  class="label label-info"
                  for="dataset-checkbox-30"
                  title="Mark dataset for download"
                >
                  <i
                    class="fa fa-download"
                  />
                  <input
                    class="dataset_checkbox ml-2"
                    id="dataset-checkbox-30"
                    name="dataset"
                    style="display: inline;"
                    type="checkbox"
                    value="30"
                  />
                </label>
              </div>
              <div
                class="float-left"
                style="margin-right: 10px;"
              />
              <div
                class="float-right"
                style="text-align: right;"
              >
                <p>
                  <span
                    class="badge badge-info"
                    title="1 file"
                  >
                    <i
                      class="fa fa-file"
                    />
                     
                    1
                  </span>
                </p>
                <p>
                  <span
                    class="badge badge-info"
                    title="Dataset size is"
                  >
                    342.35 KB
                  </span>
                </p>
              </div>
              <div
                class="float-left ds-thumb"
                style="margin-right: 10px;"
              >
                <canvas
                  style="width: 100px;"
                />
              </div>
              <div
                class="float-left"
              >
                <p
                  class="mb-2"
                >
                  <a
                    class="dataset-link"
                    href="/dataset/30"
                  >
                    New Dataset
                  </a>
                </p>
                
                <div
                  style="clear: both;"
                />
              </div>
            </div>
          </li>
        </div>
      </ul>
    </div>
  </div>
  <div
    id="experiment-tab-transfer-datasets"
  >
    <div
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="share-list"
    >
      <form
        class="form-horizontal"
        id="other-experiment-selection"
      >
        <fieldset>
          <div
            class="form-group"
          >
            <label
              class="col-md-2 col-form-label"
              for="input01"
            >
              Experiment
            </label>
            <div
              class="col-md-10"
            >
              <select
                class="form-control"
                name="experiment_id"
              >
                <option
                  value="34"
                >
                  My Test Exp
                </option>
                <option
                  value="38"
                >
                  My Test Exp 2
                </option>
              </select>
            </div>
          </div>
        </fieldset>
      </form>
      <p
        class="help-text"
      >
        <strong>
          Instructons:
        </strong>
         
        Using the above list, select an experiment to copy from, then drag datasets to the right to associate them with the current experiment.
      </p>
      <ul
        class="datasets thumbnails"
      >
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="share-list_26"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="share-list_26"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <li
            class="dataset card mb-2"
          >
            <div
              class="card-body"
            >
              <div
                class="float-left"
                style="margin-right: 10px;"
              >
                <div
                  style="display: flex; text-align: center; font-size: 32px;"
                >
                  <i
                    class="fa fa-folder-open o-6"
                  />
                </div>
                <br />
                
              </div>
              <div
                class="float-left"
                style="margin-right: 10px;"
              />
              <div
                class="float-right"
                style="text-align: right;"
              >
                <p>
                  <span
                    class="badge badge-info"
                    title="1 file"
                  >
                    <i
                      class="fa fa-file"
                    />
                     
                    1
                  </span>
                </p>
                <p>
                  <span
                    class="badge badge-info"
                    title="Dataset size is"
                  >
                    160.64 KB
                  </span>
                </p>
              </div>
              <div
                class="float-left ds-thumb"
                style="margin-right: 10px;"
              >
                <canvas
                  style="width: 100px;"
                />
              </div>
              <div
                class="float-left"
              >
                <p
                  class="mb-2"
                >
                  <a
                    class="dataset-link"
                    href="/dataset/26"
                  >
                    My demo dataset 25062019
                  </a>
                </p>
                
                <div
                  style="clear: both;"
                />
              </div>
            </div>
          </li>
        </div>
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="share-list_30"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="share-list_30"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <li
            class="dataset card mb-2"
          >
            <div
              class="card-body"
            >
              <div
                class="float-left"
                style="margin-right: 10px;"
              >
                <div
                  style="display: flex; text-align: center; font-size: 32px;"
                >
                  <i
                    class="fa fa-folder-open o-6"
                  />
                </div>
                <br />
                
              </div>
              <div
                class="float-left"
                style="margin-right: 10px;"
              />
              <div
                class="float-right"
                style="text-align: right;"
              >
                <p>
                  <span
                    class="badge badge-info"
                    title="1 file"
                  >
                    <i
                      class="fa fa-file"
                    />
                     
                    1
                  </span>
                </p>
                <p>
                  <span
                    class="badge badge-info"
                    title="Dataset size is"
                  >
                    342.35 KB
                  </span>
                </p>
              </div>
              <div
                class="float-left ds-thumb"
                style="margin-right: 10px;"
              >
                <canvas
                  style="width: 100px;"
                />
              </div>
              <div
                class="float-left"
              >
                <p
                  class="mb-2"
                >
                  <a
                    class="dataset-link"
                    href="/dataset/30"
                  >
                    New Dataset
                  </a>
                </p>
                
                <div
                  style="clear: both;"
                />
              </div>
            </div>
          </li>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
